beetsplug/discogs.py

```diff
@@ -315,30 +339,95 @@ def get_tracks(self, tracklist):
+                # Merge the subtracks, pick a title, and append the new track.
+                track = subtracks[0].copy()
+                track['title'] = ' / '.join([t['title'] for t in subtracks])
+                tracklist.append(track)
+
+        # Pre-process the tracklist, trying to identify subtracks.
+        subtracks = []
+        tracklist = []
+        prev_subindex = ''
+        for track in raw_tracklist:
+            if track['position']:
+                _, _, subindex = self.get_track_index(track['position'])
+                if subindex:
+                    if subindex.rjust(len(raw_tracklist)) > prev_subindex:
+                        # Subtrack still part of the current main track.
+                        subtracks.append(track)
+                    else:
+                        # Subtrack part of a new group (..., 1.3, *2.1*, ...).
+                        add_merged_subtracks(tracklist, subtracks)
+                        subtracks = [track]
+                    prev_subindex = subindex.rjust(len(raw_tracklist))
+                else:
+                    # Regular track.
+                    if subtracks:
+                        add_merged_subtracks(tracklist, subtracks)
+                        subtracks = []
+                        prev_subindex = ''
+                    tracklist.append(track)
+            else:
+                # Index track.
+                if subtracks:
+                    add_merged_subtracks(tracklist, subtracks)
+                    subtracks = []
+                    prev_subindex = ''
+                tracklist.append(track)
+
+        # Merge and add the remaining subtracks, if any.
+        if subtracks:
+            add_merged_subtracks(tracklist, subtracks)
+
+        return tracklist
+
     def get_track_info(self, track, index):
         """Returns a TrackInfo object for a discogs track.
         """
         title = track['title']
         track_id = None
-        medium, medium_index = self.get_track_index(track['position'])
+        medium, medium_index, _ = self.get_track_index(track['position'])
         artist, artist_id = self.get_artist(track.get('artists', []))
         length = self.get_track_length(track['duration'])
         return TrackInfo(title, track_id, artist, artist_id, length, index,
                          medium, medium_index, artist_sort=None,
                          disctitle=None, artist_credit=None)
 
     def get_track_index(self, position):
-        """Returns the medium and medium index for a discogs track position.
+        """Returns the medium, medium index and subtrack index for a discogs
+        track position.
         """
+        # TODO: revise comment
         # medium_index is a number at the end of position. medium is everything
         # else. E.g. (A)(1), (Side A, Track )(1), (A)(), ()(1), etc.
-        match = re.match(r'^(.*?)(\d*)$', position.upper())
+        match = re.match(r'^(.*?)(\d*?)(\.[\w]+|[A-Z]+)?$', position.upper())
```

Comment:
This was derived from the information at 12.2.9 on the [discogs guidelines](https://www.discogs.com/help/submission-guidelines-release-trk.html#Position) - however, 12.2.2 worries me a bit:

> 12.2.2. You can enter the positions from release, or using the Discogs standard positions (see below). The positions from the release are preferred.

Some extra checks for non-standard positions would probably be needed!